fem/hyperbolic_conservation_laws.hpp

```diff
@@ -0,0 +1,779 @@
+                     const Vector &fluxN1, const Vector &fluxN2,
+                     const double maxE, const Vector &nor, Vector &flux) = 0;
+   virtual ~RiemannSolver() = default;
+};
+
+/**
+ * @brief Abstract hyperbolic form integrator, (F(u, x), ∇v) and (F̂(u±, x, n))
+ *
+ */
+class HyperbolicFormIntegrator : public NonlinearFormIntegrator
+{
+private:
+   const int num_equations;  // the number of equations
+   // The maximum characterstic speed, updated during element/face vector assembly
+   double max_char_speed;
+   const int IntOrderOffset;  // 2*p + IntOrderOffset will be used for quadrature
+   RiemannSolver *rsolver;    // Numerical flux that maps F(u±,x) to hat(F)
+#ifndef MFEM_THREAD_SAFE
+   // Local storages for element integration
+   Vector shape;              // shape function value at an integration point
+   Vector state;              // state value at an integration point
+   DenseMatrix flux;          // flux value at an integration point
+   DenseMatrix dshape;  // derivative of shape function at an integration point
+
+   Vector shape1;  // shape function value at an integration point - first elem
+   Vector shape2;  // shape function value at an integration point - second elem
+   Vector state1;  // state value at an integration point - first elem
+   Vector state2;  // state value at an integration point - second elem
+   Vector fluxN1;  // flux dot n value at an integration point - first elem
+   Vector fluxN2;  // flux dot n value at an integration point - second elem
+   Vector nor;     // normal vector (usually not a unit vector)
+   Vector fluxN;   // hat(F)(u,x)
+#endif
+
+public:
+   /**
+    * @brief Compute flux F(u, x) for given state u and physical point x
+    *
+    * @param[in] state value of state at the current integration point
+    * @param[in] Tr Transformation to find physical location x of the current
+    * integration point
+    * @param[out] flux F(u, x)
+    * @return double maximum characteristic speed
+    *
+    * @note One can put assertion in here to detect non-physical solution
+    */
+   virtual double ComputeFlux(const Vector &state, ElementTransformation &Tr,
+                              DenseMatrix &flux) = 0;
+   /**
+    * @brief Compute normal flux. Optionally overloadded in the
+    * derived class to avoid creating full dense matrix for flux.
+    *
+    * @param[in] U U value at the current integration point
+    * @param[in] normal normal vector (usually not a unit vector)
+    * @param[in] Tr element transformation
+    * @param[out] FUdotN normal flux from the given element at the current
+    * integration point
+    * @return double maximum characteristic velocity
+    */
+   virtual double ComputeFluxDotN(const Vector &U, const Vector &normal,
+                                  ElementTransformation &Tr, Vector &FUdotN)
+   {
+      double val = ComputeFlux(U, Tr, flux);
+      flux.Mult(normal, FUdotN);
+      return val;
+   }
```

Comment:
The implementation of this (and other member functions in this file) can be moved to the cpp file.